src/ucp/core/ucp_ep.c

```diff
@@ -225,30 +225,42 @@ ucp_ep_local_disconnect_progress_remove_filter(const ucs_callbackq_elem_t *elem,
     return 1;
 }
 
-void ucp_ep_destroy_base(ucp_ep_h ep)
+static UCS_F_NOINLINE void ucp_ep_remove_progress_callbacks(ucp_ep_h ep)
 {
     ucp_worker_h worker = ep->worker;
-
-    if (--ep->ref_cnt != 0) {
-        return;
-    }
+    ucp_lane_index_t lane;
 
     /* Remove pending slow-path functions after all UCT EP lanes are destroyed,
      * because cleanup lanes purges all outstanding operation and purged
      * operations could add callbacks on the progress */
+    ucs_assert(ep->ref_cnt == 0);
+    if (ep->cfg_index != UCP_WORKER_CFG_INDEX_NULL) {
+        for (lane = 0; lane < ucp_ep_num_lanes(ep); ++lane) {
+            ucs_assert((ep->uct_eps[lane] == NULL) ||
+                       ucp_is_uct_ep_failed(ep->uct_eps[lane]));
+        }
+    }
+
+    ucs_callbackq_remove_if(&ep->worker->uct->progress_q,
+                            ucp_wireup_msg_ack_cb_pred, ep);
 
-    /* Remove pending slow-path progress in case it wasn't removed yet */
     ucs_callbackq_remove_if(&worker->uct->progress_q,
                             ucp_worker_err_handle_remove_filter, ep);
 
-    /* Remove pending slow-path function if it wasn't removed yet */
     ucs_callbackq_remove_if(&worker->uct->progress_q,
                             ucp_listener_accept_cb_remove_filter, ep);
 
-    /* Remove pending slow-path disconnect progress if it wasn't removed yet */
     ucs_callbackq_remove_if(&worker->uct->progress_q,
                             ucp_ep_local_disconnect_progress_remove_filter, ep);
+}
+
+void ucp_ep_destroy_base(ucp_ep_h ep)
+{
+    if (--ep->ref_cnt != 0) {
+        return;
+    }
 
+    ucp_ep_remove_progress_callbacks(ep);
```

Comment:
maybe do it in ucp_ep_delete()?

Reply:
but we call `ucp_ep_destroy_base()` from discarding

Reply:
didn't we already call ucp_ep_delete? or the issue is that more callbacks could have been added?

Reply:
okay